ink-wrapper-types/src/drink/client.rs

```diff
@@ -0,0 +1,164 @@
+
+        Ok(ContractInstantiateResult {
+            gas_consumed: instantiate_contract_result.gas_consumed,
+            gas_required: instantiate_contract_result.gas_required,
+            result: contract_address,
+            events,
+            reverted: false,
+        })
+    }
+
+    fn execute<T: scale::Decode + Send + std::fmt::Debug>(
+        &mut self,
+        call: ExecCall<T>,
+    ) -> Result<ContractExecResult<T>, Error> {
+        let actor = self.get_actor();
+        let gas_limit = self.get_gas_limit();
+        let contract_address = (*AsRef::<[u8; 32]>::as_ref(&call.account_id)).into();
+
+        call_contract(
+            actor,
+            gas_limit,
+            self.sandbox(),
+            contract_address,
+            call.value,
+            call.data,
+        )
+    }
+
+    fn query<T: scale::Decode + Send + std::fmt::Debug>(
+        &mut self,
+        call: ReadCall<T>,
+    ) -> Result<ContractReadResult<T>, Error> {
+        let actor = self.get_actor();
+        let gas_limit = self.get_gas_limit();
+        let contract_address = (*AsRef::<[u8; 32]>::as_ref(&call.account_id)).into();
+
+        self.sandbox().dry_run(|sandbox| {
+            call_contract(actor, gas_limit, sandbox, contract_address, 0, call.data)
+        })
+    }
+}
+
+fn call_contract<T: scale::Decode + Send + std::fmt::Debug>(
+    actor: MinimalRuntimeAccount,
+    gas_limit: Weight,
+    sandbox: &mut drink::Sandbox<MinimalRuntime>,
+    address: MinimalRuntimeAccount,
+    value: u128,
+    data: Vec<u8>,
+) -> Result<ContractResult<T>, Error> {
+    // Reset events to make sure we don't have any events from previous calls.
+    sandbox.reset_events();
+
+    let result = sandbox.call_contract(
+        address,
+        value,
+        data,
+        actor,
+        gas_limit,
+        None,
+        pallet_contracts::Determinism::Enforced,
+    );
+
+    let message_result: T = match &result.result {
+        // NOTE to reviewers: we're not checking if the contract call reverted here.
+        // We will most probably return an error from the contract call anyway.
+        Ok(exec_result) => {
+            let encoded = exec_result.data.clone();
+
+            T::decode(&mut encoded.as_slice()).map_err(|err| {
+                Error::DecodingError(format!(
+                    "Failed to decode the result of calling a contract: {err:?}",
+                ))
+            })
+        }
+        Err(err) => Err(Error::CallFailed(*err)),
+    }?;
+
+    let events = extract_events(&result.events);
+
+    Ok(ContractResult {
+        gas_consumed: result.gas_consumed,
+        gas_required: result.gas_required,
+        result: message_result,
+        events,
+        reverted: result
+            .result
+            .map(|exec_result| exec_result.did_revert())
+            .expect(
+            "If `result.result` was `err`, we should have returned `Err` from the whole function.",
```

Comment:
Question: is this formatted correctly? Looks weird, but we should have a formatting check on CI, so I assume it is?

Reply:
Rerunning ```cargo fmt --all -- --all-targets``` does not change anything for me 🤔 

I think `cargo fmt` does not work well every time. Especially in macros. Maybe this is another use case?

Reply:
I've rearranged the code a bit, should be tad better to read.